portal-web/docroot/html/common/themes/top_js.jspf

```diff
@@ -152,6 +152,11 @@
 					}
 
 					return value;
+				},
+				put: function(items) {
```

Comment:
The only caveat from my side is this method: not the method itself, but the fact that we are promoting the cache to the category of official API and we don't have any way to invalidate it.

Also, the method proposed here would prevent us from creating a garbage collector (think of an [LRU cache strategy eviction strategy](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_Recently_Used_(LRU)), for example) because if we remove keys added with put they won't be restored.

We should give that a thought, not to fix it now, but to make sure we are not doing anything that will limit us in the future.

Other than that I think the solution is good :+1:

Reply:
> We should give that a thought, not to fix it now, but to make sure we are not doing anything that will limit us in the future.

@izaera couldn't this be handled by updating this to have 2 caches? ie: `Liferay.Language._cache` and `Liferay.Language._internal_cache`. We could keep using `Liferay.Language._cache` for `put` but for the keys that Liferay adds to the cache that we want to be able to invalidate we would add them to `Liferay.Language._internal_cache`. We would have to update `get` to return keys from both caches, but then we could invalidate `Liferay.Language._internal_cache` without worrying about removing anything added by `put`.

Reply:
Yeah, that would be an option. 

However I would prefer a more resilient/automatic mechanism because your proposal assumes that users of `put` won't put too much stuff in the cache, but that may not be the case and, if we don't provide a way to manage the cache, they won't be able to clean it either.

It may be impossible to fix, in any case, because if we remove things from cache and then they are needed that would imply an async call to the server, which would defeat the purpose of the solution (leaving `Liferay.Language.get` as a synchronous API) :shrug:.

Additionally, all JS using the labels is not evicted from the browser JS engine either, so maybe this cache wouldn't be the biggest problem in a huge overload scenario... I don't know. I would have to think about it carefully. 

Just wanted to highlight that creating a cache without no clear invalidation/eviction strategy in mind tipically leads to problems in the future so better be careful.

Reply:
Now I'm thinking that I may have the same problem in the internal solution I merged :flushed: (though I swear I gave a thought to the cache eviction strategy and convinced myself that it was feasible :thinking: ).

Reply:
If a cache issue emerges it will ultimately mean there was a) great adopting of this new pattern and b) many CX using it concurrently on the page; both of which would already be huge wins.

Reply:
Also, @dsanz asked me last Friday if after each SPA navigation this cache would be cleared and that's the case, so that may be enough to keep the cache bounded in size.

Reply:
When that SPA navigation occurs? Because if it might be often, what would be the purpose of the cache if it's never really used?

Reply:
I'd like to bring the language override scenario. If a key that is cached has just changed its value, would the cache be updated?

Reply:
> When that SPA navigation occurs? Because if it might be often, what would be the purpose of the cache if it's never really used?

The purpose of the cache is to store the translations so that when `Liferay.Language.get` is invoked it can return the translation without any need to request it from the server. If it needed to go to the server then `Liferay.Language.get` would need to return a `Promise` which is not what the current contract specifies.

So more than a cache is an "pre-fetch store" or whatever you want to name it :shrug: .

Reply:
> If a key that is cached has just changed its value, would the cache be updated?

Yes, after a navigation occurs or when the user refreshes the page. This is how it works now and the only sane way to work because otherwise you would need to push updates from the server to all clients or make the clients poll the server before using any translation.

Reply:
>The purpose of the cache is to store the translations so that when Liferay.Language.get is invoked it can return the translation without any need to request it from the server.
I get it's a pre-request cache. But that's the issue. How would the frontend cache know that the language override cache has changed?

Reply:
>Turns out I was wrong and the cache is not removed with SPA navigations (just with page refresh) 

@izaera  So consider a user is at the home page with a custom app that contains a button using the 'my-cool-button' language key. If they go to the language override page, change its value from `my cool button` to `my button` and go back to the home page, without refreshing any page, just normal navigation, then the translations wouldn't be updated?

Reply:
what @izaera is trying to explain is that the frontend does not need to care if the translation was updated since on next page render the cache is empty and will be repopulated (using the updated value retrieved from the api call).

The point is that "_cache" is not really a "cache" in the pure sense but rather a _pre-fetch_ store.

Reply:
@rotty3000 that's weird... what is the point of having something called `cache` that is not a cache and is cleared in every page render?

Reply:
Don't forget that the assumption is that `Liferay.Language.put(await fetchTranslations(keys));` will be called, for example, in the [`HTMLElement.connectedCallback()`](https://developer.mozilla.org/en-US/docs/Web/API/Web_components#connectedcallback) of custom element CX.

Reply:
it's cached during the entire lifetime of the page render... which might be a long time, so it's still a cache from that perspective.

Reply:
> If they go to the language override page, change its value from my cool button to my button and go back to the home page, without refreshing any page, just normal navigation, then the translations wouldn't be updated?

This would in fact update the translation in the rendered page since you navigated away and came back.

Reply:
Never heard of this kind of caching. Does it really have that much of an impact?

Reply:
This is not semantically different than updating anything else in the portal. like editing the page and moving a widget. Others need to update the page to see the change. Pages are not updated in real time.

Reply:
hmm okay, makes sense. Thanks @rotty3000

Reply:
Changing a key is infrequent and only by admins. I don't think we have to worry about solving that problem. We don't need perfect being the enemy of the good enough.

Reply:
FYI, the `Liferay.Language._cache` is literally nothing more than a empty JavaScript object (i.e. https://github.com/liferay-platform-experience/liferay-portal/pull/786/files#diff-85bb8302271a11c1e5ab780bbea0be403a73665d742def70c0d817f8d43e6713R137-R144). There is literally no logic around it. It's effectively an empty map.

The problem today is that if you call `Liferay.Language.get('key-x')` only `key-x` is returned. There is no fetch to Liferay to return the correct translation AND we cannot make it do that because the API is sync and you cannot make a sync api async (which would be required to do a `fetch`).

Now, we don't actually want to make `Liferay.Language.get('key-x')` fetch each time even if we could make it work because then we would have N fetches on every page render for all those translations which would be a huge network cost.

So, a simple solution is, since the existing "cache" is not really used, AND since we want to make `Liferay.Language.get` work, but we don't want all those network calls, we can populate this temporary cache ONCE just before render with a single api call which has all the expected keys. This will fill the "cache" so that all the `Liferay.Language.get` work.

Does this make sense?

Reply:
Now I get it right. Sorry, I got confused with the cache being pre-fetch, new concept to me.

Reply:
@rotty3000 If the cache is pre-fetch and we get a new one pre page rendering, why is @izaera concerned about having a garbage colletor?

> Also, the method proposed here would prevent us from creating a garbage collector (think of an [LRU cache strategy eviction strategy](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_Recently_Used_(LRU)), for example) because if we remove keys added with put they won't be restored.

Reply:
🤷🏻 I personally don't see the need but I could be missing something.

Reply:
One thing to keep in mind is that there may be multiple widgets on the page that each make a call to
```javascript
Liferay.Language.put(await fetchTranslations(keys));
```
and maybe that is what @izaera is referring too. I still don't see the need to optimize that just yet.

My basic assumption is that there will still be far less network traffic (than with individual translation requests), but not zero. And there may even be some duplication of work.

For example:

suppose in app1 we have:
```javascript
Liferay.Language.put(await fetchTranslations(['red', 'green']));
```
and in app2, we have:
```javascript
Liferay.Language.put(await fetchTranslations(['green', 'blue']));
```
this is two network calls, duplicating the getting of translation for the key 'green' and putting into the cache. But that isn't really a huge issue because both apps will see the correct translation even though the second request is doing a put over an existing value... both translations came from the portal so this not a big deal.

### Future Consideration
We might consider later to integrate the developer requesting of keys into the CX framework itself such that a CX tells the portal which keys it uses so that on the server side, before rendering any of the page at all, can collect all the requested keys and fill "_cache". But that is also a much more elaborate solution that requires quite a bit of orchestration.

However, That doesn't void the interesting need to have an API call to aggregate translations (in the meantime or afterward even), and so the meat of this PR doesn't change even when considering that possibility.

Reply:
>@rotty3000 If the cache is pre-fetch and we get a new one pre page rendering, why is @izaera concerned about having a garbage colletor?

We don't get a new one with every page render. 

**Under SPA**: we get a new one with every refresh (which requires user to press F5, so not a regular workflow). When you navigate using SPA the cache is maintained (see https://github.com/liferay/liferay-portal/commit/d32bc3523f25d4afc55072f71c3b733beb31687e). 

**Without SPA**: we get a new one with every page navigation (because it implies a refresh).

My concern is what would happen if you navigate for one week under SPA and translations keep growing. The cache could end up being huge. Just that.

On the other hand, if the cache is huge, the .js files associated to translations (because remember that translations are added by .js modules, they are not something that appear out of the blue) will probably be even bigger than the translations. I'm talking in terms of RAM. If you think this is futile simply open an LPD in JIRA and leave it there for 3-4 days, then try to interact with it at normal speeds :smiling_imp: .

In any case, since the cache is user managed, if we need cache management in the future, we can simply add a `delete` method and leave it up to the users. This may be better because they would know better what can be deleted from cache and when. :shrug:

Reply:
Btw, what I mean by "rendering" is the lifecycle of `HTMLElement` in JavaScript, though of course there are other use cases for using `Liferay.Language.get('..')` that custom elements.

I suggest we should have an observation period to see how things behave.